opal/rocm/common_rocm.c

```diff
@@ -0,0 +1,319 @@
+/* -*- Mode: C; c-basic-offset:4 ; indent-tabs-mode:nil -*- */
+/*
+ * Copyright (c) 2022 Advanced Micro Devices, Inc. All rights reserved.
+ *
+ * $COPYRIGHT$
+ *
+ * Additional copyrights may follow
+ *
+ * $HEADER$
+ */
+
+#include <stdio.h>
+#include <dlfcn.h>
+#include "opal_config.h"
+#include "common_rocm.h"
+#include "opal/mca/dl/base/base.h"
+#include "opal/runtime/opal_params.h"
+
+static void* hip_handle = NULL;
+static hipFunctionTable_t hip_funcs;
+
+int mca_common_rocm_memcpy_async = 1;
+int mca_common_rocm_verbose = 0;
+size_t mca_common_rocm_memcpy_limit=1024;
+
+static bool rocm_initialized = false;
+static hipStream_t common_rocm_MemcpyStream;
+static opal_mutex_t common_rocm_init_lock;
+
+#define HIP_CHECK(condition)                                                 \
+{                                                                            \
+    hipError_t error = condition;                                            \
+    if(hipSuccess != error){                                                 \
+        const char* msg = hip_funcs.hipGetErrorString(error);                \
+        opal_output(0, "HIP error: %d %s file: %s line: %d\n", error, msg, __FILE__, __LINE__); \
+        return error;                                                        \
+    }                                                                        \
+}
+
+#define HIP_CHECK_RETNULL(condition)                                         \
+{                                                                            \
+    hipError_t error = condition;                                            \
+    if(hipSuccess != error){                                                 \
+        const char* msg = hip_funcs.hipGetErrorString(error);                \
+        opal_output(0, "HIP error: %d %s file: %s line: %d\n", error, msg, __FILE__, __LINE__); \
+        return NULL;                                                         \
+    }                                                                        \
+}
+
+static int hip_dl_init(void)
+{
+    char *str;
+    void *ptr;
+    int ret  = opal_dl_open("libamdhip64.so", false, false, &hip_handle, &str);
+    if (OPAL_SUCCESS != ret) {
+        opal_output(0, "Unable to open libamdhip64.so\n");
+        return OPAL_ERROR;
+    }
+
+    ret = opal_dl_lookup(hip_handle, "hipMalloc", &ptr, &str);
+    if (OPAL_SUCCESS != ret) {
+        opal_output(0, "Failed to find hipMalloc\n");
+        dlclose(hip_handle);
+        return OPAL_ERROR;
+    }
+    *(void **)(&hip_funcs.hipMalloc) = ptr;
```

Comment:
We had a chat about this off-line with Brian. After some discussion we agreed that I convert the code to use something like:

typedef hipError_t(*hipMalloc_t)(void **, size_t);
hip_funcs.hipMalloc = (hipMalloc_t)ptr;

understanding that this will generate a warning with the gcc compiler using the -pedantic flag. This has been now done and implemented. I will check whether there is a way to tell gcc to ignore this warning (but I haven't found it so far).